nighthawk/source/client/output_formatter_impl.cc

```diff
@@ -0,0 +1,86 @@
+#include "nighthawk/source/client/output_formatter_impl.h"
+
+#include <chrono>
+#include <sstream>
+
+#include "common/protobuf/utility.h"
+
+namespace Nighthawk {
+namespace Client {
+
+OutputFormatterImpl::OutputFormatterImpl(Envoy::TimeSource& time_source, const Options& options,
+                                         const std::vector<StatisticPtr>& merged_statistics,
+                                         const std::map<std::string, uint64_t>& merged_counters)
+    : time_source_(time_source), options_(options), merged_statistics_(merged_statistics),
+      merged_counters_(merged_counters) {}
+
+ConsoleOutputFormatterImpl::ConsoleOutputFormatterImpl(
+    Envoy::TimeSource& time_source, const Options& options,
+    const std::vector<StatisticPtr>& merged_statistics,
+    const std::map<std::string, uint64_t>& merged_counters)
+    : OutputFormatterImpl(time_source, options, merged_statistics, merged_counters) {}
+
+std::string ConsoleOutputFormatterImpl::toString() const {
+  std::stringstream s;
+
+  // TODO(oschaaf): echo non-default options to CLI output?
+  s << "Merged statistics:\n";
+  for (auto& statistic : merged_statistics_) {
+    if (statistic->count() > 0) {
+      s << fmt::format("{}: {}\n", statistic->id(), statistic->toString(), "\n");
+    }
+  }
+  s << "\nMerged counters\n";
+  for (auto counter : merged_counters_) {
+    s << fmt::format("counter {}:{}\n", counter.first, counter.second);
+  }
+  return s.str();
+}
+
+Envoy::ProtobufTypes::MessagePtr OutputFormatterImpl::toProto() const {
+  nighthawk::client::Output output;
+  output.set_allocated_options(options_.toCommandLineOptions().release());
+
+  auto ts = time_source_.systemTime().time_since_epoch();
+  auto seconds = std::chrono::duration_cast<std::chrono::seconds>(ts);
+  output.mutable_timestamp()->set_seconds(seconds.count());
+  output.mutable_timestamp()->set_nanos(
```

Comment:
There is some standard `Duration` and time conversion, see https://github.com/envoyproxy/envoy/blob/master/source/common/protobuf/utility.cc#L333 (you might need thereverse direction).

Reply:
I think this is better now